### DISCUSSION TITLE
I hit "may not be able to detach the kernel driver for this device" on MacOS

### DISCUSSION THREAD

**Discussion by real-felix** [libusb/libusb#1321] [Q&A] [answered]
Hello, I try and read the HID reports on MacOS, but I get the message saying that the kernel driver cannot be detached:

```
% sudo usbhid-dump --model=16c0:27db
Password:
libusb: info [darwin_detach_kernel_driver] no capture entitlements. may not be able to detach the kernel driver for this device
libusb: warning [darwin_transfer_status] transfer error: timed out
000:001:000:Failed to retrieve report descriptor: Operation timed out
libusb: error [darwin_reenumerate_device] darwin/reenumerate_device: timeout waiting for reenumerate
000:001:000:Failed to attach to the kernel driver: Operation timed out
libusb: warning [darwin_close] darwin_close device missing IOService
```

What does it mean? How can I debug my keyboard on MacOS?

**Comment by mcuee** [comment]
libusb is not the right libray to use for HID device under macOS, as you can not detach the kernel HID driver. 
* https://github.com/libusb/libusb/issues/1014
* https://github.com/libusb/libusb/wiki/FAQ#user-content-Does_libusb_support_USB_HID_devices

I am not sure if HIDAPI can help you or not since your device is a keyboard. You can give it a try.
https://github.com/libusb/hidapi

You may want to ask in Apple forum to get some suggestions.

**Comment by real-felix** [comment]
Alright, thank you for the answer! Never mind, I'll just boot my old linux laptop, I wanted a solution out-of-the-box: if it does not work directly, I'll not mess up with Apple stuff.

Have you considered adding a Github discussion? I feel bad writing an issue only for asking a question.

**Comment by mcuee** [comment]
We recommend using libusb-devel mailing list for technical support.

There are too few libusb developers now to be able to support github discussion.

**Reply by Zheng-Bicheng** [reply]
Some countries' users do not have the habit of using email for communication, and the time to reply to emails may be long, so github discussions are necessary.

**Comment by real-felix** [comment]
Well, it does not work either on Linux, but for a different reason. First `lsusb` displays the name as something completely different (definitely not the one I put in the firmware).

Then I get:

```
$ sudo usbhid-dump --model=16c0:27db
Starting dumping interrupt transfer stream
with 1 minute timeout.

005:004:000:Failed to set report protocol: Operation timed out
005:004:000:Failed to attach to the kernel driver: Entity not found
```

And then my keyboard does not work anymore, I have to "reboot" it.

EDIT: sorry, I did not see your last message before posting…

**Comment by mcuee** [comment]
For Linux, when your use libusb and detach the kernel driver, your keyboad function will not work. That is why you need to re-attach the kernel driver once you finish the application.

Therefore you should use HIDAPI as well under Linux.

On the othe hand, it seems your HID device is not working well. Are you developing MCU firmware for a USB keyboard? You should consider using debugging tool for your MCU and use debugger (eg: JTAG/SWD debugger or emulator), or use UART to dump the debugging info.

**Reply by real-felix** [reply, marked as answer]
> Are you developing MCU firmware for a USB keyboard?

Yeah, that's right, how could you tell :P

My keyboard works well, I use it to type right now, but I assume that I forgot to so something in the USB class implementation…Anyway, I'll follow your advice and use HIDAPI to debug my reports.

Unfortunately, I cannot use any kind of hardware debugging because I didn't build this keyboard, and the pins are not accessible. I'm afraid of breaking everything if I do something to the board.

**Comment by mcuee** [comment]
> We recommend using libusb-devel mailing list for technical support.
> 
> There are too few libusb developers now to be able to support github discussion.

In the end, I think we may add github discussion since some people may not like to use libusb-devel mailing list.

**Reply by real-felix** [reply]
I'm young, I have no clue how to use a mailing list to be honest 😅
